### PR TITLE
`index_select` for COO CUDA tensors.

### PR DESCRIPTION
**Pull request by nikitaved**
Brings a native CUDA implementation for `index_select`. Master silently converts CUDA tensors to CPU for CUDA support.

Case `nnz >> size` could be optimized similar to how https://github.com/pytorch/pytorch/pull/72710 is doing that.

Some benchmarks:
<details>

<summary>PR/torch_sparse/master</summary>

```
[------------------------------- cuda coo.index_select -------------------------------]
                                                    |   PR   |  torch_sparse  |  master
32 threads: ---------------------------------------------------------------------------
      n=10000, nnz=100, index_len=100, dim=0        |    96  |       327      |     70 
      n=10000, nnz=100, index_len=100, dim=1        |   120  |       505      |     74 
      n=10000, nnz=100, index_len=1000, dim=0       |    90  |       333      |     93 
      n=10000, nnz=100, index_len=1000, dim=1       |   120  |       499      |     98 
      n=10000, nnz=100, index_len=10000, dim=0      |    92  |       331      |    350 
      n=10000, nnz=100, index_len=10000, dim=1      |   100  |       506      |    352 
      n=100000, nnz=1000, index_len=100, dim=0      |    53  |       274      |     60 
      n=100000, nnz=1000, index_len=100, dim=1      |    90  |       368      |     71 
      n=100000, nnz=1000, index_len=1000, dim=0     |    93  |       332      |    100 
      n=100000, nnz=1000, index_len=1000, dim=1     |   130  |       501      |    140 
      n=100000, nnz=1000, index_len=10000, dim=0    |   100  |       341      |    522 
      n=100000, nnz=1000, index_len=10000, dim=1    |   130  |       530      |    549 
      n=1000000, nnz=10000, index_len=100, dim=0    |    90  |       429      |    110 
      n=1000000, nnz=10000, index_len=100, dim=1    |   296  |       810      |    355 
      n=1000000, nnz=10000, index_len=1000, dim=0   |   100  |       435      |    170 
      n=1000000, nnz=10000, index_len=1000, dim=1   |   309  |       830      |    548 
      n=1000000, nnz=10000, index_len=10000, dim=0  |   110  |       446      |    750 
      n=1000000, nnz=10000, index_len=10000, dim=1  |   310  |       830      |   1000 
      n=10, nnz=100, index_len=100, dim=0           |    90  |       333      |     74 
      n=10, nnz=100, index_len=100, dim=1           |   100  |       497      |     78 
      n=10, nnz=100, index_len=1000, dim=0          |    90  |       329      |    140 
      n=10, nnz=100, index_len=1000, dim=1          |   100  |       800      |    100 
      n=10, nnz=100, index_len=10000, dim=0         |    93  |       340      |    900 
      n=10, nnz=100, index_len=10000, dim=1         |   120  |       800      |    489 
      n=10, nnz=1000, index_len=100, dim=0          |    90  |       321      |    140 
      n=10, nnz=1000, index_len=100, dim=1          |   100  |       680      |    140 
      n=10, nnz=1000, index_len=1000, dim=0         |   110  |       349      |    670 
      n=10, nnz=1000, index_len=1000, dim=1         |   130  |       740      |    800 
      n=10, nnz=1000, index_len=10000, dim=0        |   302  |       503      |   4882 
      n=10, nnz=1000, index_len=10000, dim=1        |   325  |      2257      |   5262 
      n=10, nnz=10000, index_len=100, dim=0         |   229  |       349      |    810 
      n=10, nnz=10000, index_len=100, dim=1         |   433  |       870      |    700 
      n=10, nnz=10000, index_len=1000, dim=0        |   666  |       502      |   5581 
      n=10, nnz=10000, index_len=1000, dim=1        |   826  |      2379      |   4820 
      n=10, nnz=10000, index_len=10000, dim=0       |  2534  |      2700      |  80000 
      n=10, nnz=10000, index_len=10000, dim=1       |  2723  |     18540      |  80000 
      n=100, nnz=1000, index_len=100, dim=0         |    94  |       324      |    110 
      n=100, nnz=1000, index_len=100, dim=1         |   100  |       499      |    110 
      n=100, nnz=1000, index_len=1000, dim=0        |    96  |       337      |    150 
      n=100, nnz=1000, index_len=1000, dim=1        |   130  |       800      |    140 
      n=100, nnz=1000, index_len=10000, dim=0       |   100  |       346      |    900 
      n=100, nnz=1000, index_len=10000, dim=1       |   130  |       760      |    900 
      n=100, nnz=10000, index_len=100, dim=0        |    90  |       323      |    190 
      n=100, nnz=10000, index_len=100, dim=1        |   279  |       800      |    180 
      n=100, nnz=10000, index_len=1000, dim=0       |   110  |       339      |    781 
      n=100, nnz=10000, index_len=1000, dim=1       |   294  |       870      |    800 
      n=100, nnz=10000, index_len=10000, dim=0      |   315  |       505      |   6264 
      n=100, nnz=10000, index_len=10000, dim=1      |   497  |      2398      |   5404 
      n=1000, nnz=10000, index_len=100, dim=0       |    90  |       333      |    160 
      n=1000, nnz=10000, index_len=100, dim=1       |   279  |       635      |    150 
      n=1000, nnz=10000, index_len=1000, dim=0      |   100  |       328      |    215 
      n=1000, nnz=10000, index_len=1000, dim=1      |   287  |       810      |    207 
      n=1000, nnz=10000, index_len=10000, dim=0     |   100  |       339      |    900 
      n=1000, nnz=10000, index_len=10000, dim=1     |   291  |       880      |   1000 
      n=1000, nnz=100000, index_len=100, dim=0      |    92  |       358      |    435 
      n=1000, nnz=100000, index_len=100, dim=1      |   302  |       900      |    530 
      n=1000, nnz=100000, index_len=1000, dim=0     |   130  |       360      |   1000 
      n=1000, nnz=100000, index_len=1000, dim=1     |   329  |       930      |   1200 
      n=1000, nnz=100000, index_len=10000, dim=0    |   343  |       530      |   7000 
      n=1000, nnz=100000, index_len=10000, dim=1    |   545  |      2446      |   6100 
      n=1000, nnz=1000000, index_len=100, dim=0     |   355  |       394      |   2210 
      n=1000, nnz=1000000, index_len=100, dim=1     |  1660  |      2276      |   2674 
      n=1000, nnz=1000000, index_len=1000, dim=0    |   877  |       574      |   6700 
      n=1000, nnz=1000000, index_len=1000, dim=1    |  2449  |      3782      |   9000 
      n=1000, nnz=1000000, index_len=10000, dim=0   |  3112  |      2931      |  57000 
      n=1000, nnz=1000000, index_len=10000, dim=1   |  7340  |     20220      |  65700 

Times are in microseconds (us).

```

</details>

### CODE REVIEW

**Comment by nikitaved** [comment]
For future reference, we can significantly improve the case `nnz > n` with a straightforward CUDA implementation using all the tools that TI brings. I still believe all these kernels could be made even better with shared memory, but I am not yet sufficiently apt with CUDA magic.

**Comment by cpuhrsch** [comment]
Hm, it seems like in the `n=10, nnz=10000, index_len=1000, dim=1 ` case this actually introduces a regression? Maybe we can detect this condition and fall back to the CPU fallback.

**Comment by nikitaved** [comment]
A regression is because the algo is naive (nnz * index_len), so we can indeed cat at some value.

**Comment by nikitaved** [comment]
@cpuhrsch , I have just realized that Timer has a huge variance per testing time. I will reduce that with more time-to-test and update the benchmark to see whether the case you are talking about is indeed a regression.

**Comment by cpuhrsch** [comment]
@nikitaved - I personally find writing a one-off script [like this one](https://github.com/pytorch/pytorch/pull/76996#issuecomment-1123306850) more transparent for benchmarking. I wonder what your take on that is?

**Comment by nikitaved** [comment]
I like it! Maybe some of the features, especially relative scale, could be introduced into `torch.utils.benchmark`. CC @robieta .

File: test/test_sparse.py [hunk 1]
```diff
@@ -1068,6 +1068,7 @@ def run_test(sizes):
         # more sophisticated algos
         run_test((10, 100, 100))
 
+    @onlyCPU
```
Comment (cpuhrsch) [inline test/test_sparse.py:1071]: Are you introducing this because otherwise the test would take too long?

Comment (nikitaved) [reply test/test_sparse.py:1071]: No, this explicitly tests parallel_for in the CPU kernel.

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 2]
```diff
@@ -1243,5 +1246,213 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
...
+              [ptr_intrsc_counts_nneg_index, index_len] GPU_LAMBDA (
+                index_t dim_val, index_t idx_val, // dim_indices[j], j
+                index_t idx_idx // i
+              ) -> index_t {
+                if (dim_val == idx_val) {
+                  applyFastAtomicIndexInc(
+                    ptr_intrsc_counts_nneg_index,
+                    idx_idx,
+                    static_cast<index_t>(index_len)
+                  );
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/Indexing.cu:1364]: @cpuhrsch , the regressed case `n=10, nnz=10000, index_len=1000` really fails here. Since `n=10` we have on average about 100 threads trying to write into the same memory, and synchronizations on CUDA are pretty slow... Therefore I have this question, how likely it is that we have such a case for GPUs? Do we merge this and wait on improvement, or shall we just implement these improvements in this PR?

Comment (cpuhrsch) [reply aten/src/ATen/native/cuda/Indexing.cu:1364]: @nikitaved - I'm not sure how likely we're to hit this case, because we don't have integrations into reference models for this. Can't we go back to our CPU-fallback baseline for this case if it's easy to detect and optimize once we've integrated this function into a model and this case shows up?

Comment (nikitaved) [reply aten/src/ATen/native/cuda/Indexing.cu]: Not relevant anymore, with the newest update we are golden.

**Comment by cpuhrsch** [comment]
@ngimel - could I ask you to take a look at the usages of thrust here? I'm not too familiar with the pros and cons of that library.

File: aten/src/ATen/cuda/Atomic.cuh [hunk 3]
```diff
@@ -168,6 +168,18 @@ static inline __device__ int32_t gpuAtomicAdd(int32_t *address, int32_t val) {
   return atomicAdd(address, val);
 }
 
+static inline __device__ int32_t gpuAtomicAddWithReturn(int32_t *address, int32_t val) {
+  return atomicAdd(address, val);
+}
+
+// WARNING: int64_t gets cast to an unsigned type! Beware of potential overflows
+static inline __device__ int64_t gpuAtomicAddWithReturn(int64_t *address, int64_t val) {
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/cuda/Atomic.cuh:176]: Right, forgot to clear that up.

File: aten/src/ATen/cuda/Atomic.cuh [hunk 4]
```diff
@@ -168,6 +168,18 @@ static inline __device__ int32_t gpuAtomicAdd(int32_t *address, int32_t val) {
...
+  using cuda_int64_t = unsigned long long int;
+  return static_cast<int64_t>(
+      atomicAdd(reinterpret_cast<cuda_int64_t*>(address), static_cast<cuda_int64_t>(val))
+  );
+}
```
[the remark this answers is not in the document]
Comment (xwang233) [reply aten/src/ATen/cuda/Atomic.cuh]: Sorry I just saw this comment. We don't have CI with cuda cc < 3.5 and I think cuda cc < 3.5 is no longer supported by pytorch?

[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/cuda/Atomic.cuh]: You are right. Thank you, @xwang233 !

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 5]
```diff
@@ -1252,5 +1252,215 @@ Tensor & masked_fill__cuda(Tensor& self, const Tensor & mask, const Tensor & val
...
+      else {
+        return dim_indices.sort();
+      }
+    }();
+
+    Tensor intrsc_counts_nneg_index;
+    Tensor intrsc_first_match_nneg_index;
+    std::tie(intrsc_counts_nneg_index, intrsc_first_match_nneg_index) = [&]() -> std::tuple<Tensor, Tensor> {
+      auto intrsc_counts_nneg_index = at::zeros_like(nneg_index);
+      auto intrsc_first_match_nneg_index = at::zeros_like(nneg_index);
```
Comment (nikitaved) [inline aten/src/ATen/native/cuda/Indexing.cu:1360]: TODO @nikitaved: replace with `empty_like`.
